src/simulators/ch/ch_runner.hpp

```diff
@@ -0,0 +1,661 @@
+{
+  switch(branch) //Decomposition of the CCZ gate into Cliffords
+  {
+    case 1:
+      states[rank].CZ(control_1, control_2);
+      break;
+    case 2:
+      states[rank].CZ(control_1, target);
+      break;
+    case 3:
+      states[rank].CZ(control_2, target);
+      break;
+    case 4:
+      states[rank].CZ(control_1, control_2);
+      states[rank].CZ(control_1, target);
+      states[rank].Z(control_1);
+      break;
+    case 5:
+      states[rank].CZ(control_1, control_2);
+      states[rank].CZ(control_2, target);
+      states[rank].Z(control_2);
+      break;
+    case 6:
+      states[rank].CZ(control_1, target);
+      states[rank].CZ(control_2, target);
+      states[rank].Z(target);
+      break;
+    case 7:
+      states[rank].CZ(control_1, control_2);
+      states[rank].CZ(control_1, target);
+      states[rank].CZ(control_2, target);
+      states[rank].Z(control_1);
+      states[rank].Z(control_2);
+      states[rank].Z(target);
+      coefficients[rank] *= -1; //Additional phase
+      break;
+    default: //Identity
+      break;
+  }
+}
+
+//-------------------------------------------------------------------------
+//Measurement
+//-------------------------------------------------------------------------
+
+double Runner::NormEstimation(uint_t n_samples, AER::RngEngine &rng)
+{
+  std::vector<uint_t> adiag_1(n_samples, 0ULL);
+  std::vector<uint_t> adiag_2(n_samples, 0ULL);
+  std::vector< std::vector<uint_t> > a(n_samples, std::vector<uint_t>(n_qubits, 0ULL));
+  #pragma omp parallel for if(n_omp_threads > 1)
+  for (size_t l=0; l<n_samples; l++)
+  {
+    for (size_t i=0; i<n_qubits; i++)
+    {
+      for (size_t j=i; j<n_qubits; j++)
+      {
+          if(rng.rand() < 0.5)
+          {
+              a[l][i] |= (1ULL << j);
+              a[l][j] |= (1ULL << i);
+          }
+      }
+      adiag_1[l] |= (a[l][i] & (1ULL << i));
+      if (rng.rand() < 0.5)
+      {
+          adiag_2[l] |= (1ULL << i);
+      }
+    }
+  }
+  return ParallelNormEstimate(states, coefficients, adiag_1, adiag_2, a, n_omp_threads);
+}
+
+double Runner::NormEstimation(uint_t n_samples, std::vector<pauli_t> generators, AER::RngEngine &rng)
+{
+  apply_pauli_projector(generators);
+  return NormEstimation(n_samples, rng);
+}
+
+uint_t Runner::MetropolisEstimation(uint_t n_steps, AER::RngEngine &rng)
+{
+  InitMetropolis(rng);
+  for (uint_t i=0; i<n_steps; i++)
+  {
+    MetropolisStep(rng);
+  }
+  return x_string;
+}
+
+std::vector<uint_t> Runner::MetropolisEstimation(uint_t n_steps, uint_t n_shots, AER::RngEngine &rng)
+{
+  std::vector<uint_t> shots(n_shots, zer);
+  shots[0] = MetropolisEstimation(n_steps, rng);
+  for (uint_t i=1; i<n_shots; i++)
+  {
+    MetropolisStep(rng);
+    shots[i] = x_string;
+  }
+  return shots;
+}
+
+void Runner::InitMetropolis(AER::RngEngine &rng)
+{
+  accept = 0;
+  //Random initial x_string from RngEngine
+  uint_t max = (1ULL<<n_qubits) - 1;
```

Comment:
I'm not a big fun of this type we have created: `uint_t` which is really a 64 bits unsigned integer... not your fault though, we have been using it all over the code base.